Project 9a - Pull Request

jackmalzone opened the pull request:
Aside from API calls and a refactoring overhaul, this pull request addresses several relevant issues and enhancements for Project 9. Below are the detailed changes:

**Form Submission Loading State:**

- Implemented renderLoading method to provide visual feedback during form submissions for profile editing, adding cards, and updating avatars.

**Card Like/Unlike Functionality:**

- Fixed the like/unlike button functionality by ensuring correct API calls and handling responses properly.
- Added error handling to manage cases where the server response might be missing expected properties.

**Card Deletion with Confirmation Popup:**

- Integrated a confirmation popup before deleting a card to ensure user confirmation.
- Ensured the confirmation popup is displayed correctly and the card is deleted upon confirmation.

**API Integration for Initial Cards and User Info:**

- Removed hardcoded initial cards and replaced them with cards fetched from the API.
- Ensured the application fetches and displays user information and cards correctly from the server.

**Code Cleanup:**

- Removed unnecessary console logs for a cleaner codebase.
- Ensured all API responses are logged appropriately for debugging purposes.

jackmalzone committed:
create api class, instantiate in userinfo

jackmalzone committed:
instantiate api for initialcards, refactor api to be more DRY

jackmalzone committed:
refactor Card.js

jackmalzone committed:
refactor index.js

jackmalzone committed:
add edit avatar button functionality

jackmalzone committed:
debugging

jackmalzone committed:
debug hell

jackmalzone committed:
like hell

jackmalzone committed:
Update index.js

jackmalzone committed:
hope lost

jackmalzone committed:
fix save profile sr-only span from being appended to the save button, adjust edit avatar svg size

jackmalzone committed:
fix delete card flow

jackmalzone committed:
fix like/unlike functionality

jackmalzone committed:
remove various console.logs

jackmalzone committed:
Merge pull request #1 from jackmalzone/project-9

Project 9

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,89 @@
+    return Promise.reject(`Error: ${res.status}`);
+  }
+
+  async _request(endpoint, options = {}) {
```
Nice method `_request`

gennady-bars commented on src/components/Card.js:
```diff
@@ -1,10 +1,26 @@
+import api from "../utils/Api.js";
+import { cardList } from "../utils/constants";
```
Please, delete these imports

From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.

gennady-bars commented on src/components/Card.js:
```diff
@@ -17,25 +33,47 @@ export default class Card {
-      .querySelector(".card__like-button")
-      .classList.toggle("card__like-button_active");
+  async _handleLikeButton() {
+    const likeButton = this._cardElement.querySelector(".card__like-button");
```
**Could be improved**

All elements found **every time** you call a method should be defined as **class fields** in the constructor, `_setEventListeners` (or in `getView`) only **1 time** in order not to waste resources on searching them again and again.

`Class fields` (the so-called **class variables** with `this`) are accessible inside any class method. You shouldn’t pass **class fields** into the call of any method or search them again because you can use class fields inside all methods.
There can be `this._inputs`, `this._submitButton`, `this._cardImage`, `this._form`, `this._likeButton`, and so on.

```js
  getView() {
    this._element = this._getTemplate();
    this._likeButton = this._element.querySelector(".card__like-button");
    this._cardImage = this._element.querySelector(".card__image");
    this._trashButton = this._element.querySelector(".card__trash-button");

```

gennady-bars commented on src/components/Card.js:
```diff
@@ -45,15 +83,39 @@ export default class Card {
   }
 }
+
+export async function renderInitialCards(
```
`renderInitialCards` should be defined in `index.js` because `Card.js` is only for defining class `Card` (not functions)

gennady-bars commented on src/components/Card.js:
```diff
@@ -45,15 +83,39 @@ export default class Card {
+) {
+  try {
+    const result = await api.getInitialCards();
+    result.forEach((data) => {
```
You need to use method `renderItems` of class `Section` for rendering the array of cards

gennady-bars commented on src/components/PopupWithConfirmation.js:
```diff
@@ -0,0 +1,22 @@
+  }
+
+  open(handleConfirmation) {
+    this._handleConfirmation = handleConfirmation;
```
NIce function updating

gennady-bars commented on src/components/PopupWithConfirmation.js:
```diff
@@ -0,0 +1,22 @@
+    this._popupForm.addEventListener("submit", (evt) => {
+      evt.preventDefault();
+      this._handleConfirmation();
+      this.close();
```
`this.close();`

Please, delete it

You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -8,26 +8,37 @@ export default class PopupWithForm extends Popup {
-      this.close();
-    };
-    this._popupForm.addEventListener("submit", this._submitHandler);
+  renderLoading(isLoading) {
```
Nice method `renderLoading`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -8,26 +8,37 @@ export default class PopupWithForm extends Popup {
+        .finally(() => this.renderLoading(false));
+    });
+  }
 
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // Here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the profile in `index.js`

gennady-bars commented on src/components/UserInfo.js:
```diff
@@ -1,18 +1,46 @@
+import api from "../utils/Api.js";
```
* From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.

gennady-bars replied:
Please, delete it


* From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,38 +35,98 @@ const profileAddFormValidator = new FormValidator(
   profileAddForm
 );
 
+const profileAvatarFormValidator = new FormValidator(
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object: `formValidators`.  Then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // Here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // Here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -8,26 +8,37 @@ export default class PopupWithForm extends Popup {
       evt.preventDefault();
+      this.renderLoading(true);
+      this._handleFormSubmit(this._getInputValues())
+        .then(() => this.close())
```
You duplicate the logics in `index.js` as well. You should decide where to have `then`, `catch`, `finally` and remove redundant promise chains

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,38 +35,98 @@ const profileAddFormValidator = new FormValidator(
+
+async function handleEditFormSubmit(data) {
+  try {
+    profileEditPopup.renderLoading(true);
```
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -119,7 +198,26 @@ profileAddButton.addEventListener("click", () => {
+    userInfo.setUserInfo(userInfoData);
+
+    const initialCardsData = await api.getInitialCards();
+    initialCardsData.forEach((data) => {
```
* You need to use method `renderItems` of class `Section` for rendering the array of cards

gennady-bars commented on src/pages/index.js:
```diff
@@ -119,7 +198,26 @@ profileAddButton.addEventListener("click", () => {
+});
+
+// Initial render
+cardSection.renderItems();
```
`cardSection.renderItems();` is not needed here because it's an empty call

It should work like that in line 213
```js
cardSection.renderItems(initialCardsData);

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,38 +35,98 @@ const profileAddFormValidator = new FormValidator(
+  } catch (err) {
+    console.error("Error updating avatar:", err);
+  } finally {
+    profileAvatarPopup.renderLoading(false);
```
Good job that you return the button text in `finally`

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Jack!

Great job on submitting your project! This is a significant milestone, and you should be proud of your progress.

**Parts I really liked in your project:**

* Nice method `_request`
* Nice method `renderLoading`
* Good job that you `catch` possible errors at the end of server requests
* Good job that you return the button text in `finally`



There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code  and the comments below:
* From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.
* You need to use method `renderItems` of class `Section` for rendering the array of cards
* You need to close modals only in blocks `then` after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well



There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all   comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

jackmalzone committed:
correct code, begin debugging corrected code

jackmalzone committed:
DRY Refactoring, Fixing, Debugging

gennady-bars commented on src/components/Card.js:
```diff
@@ -1,59 +1,77 @@
+import api from "../utils/Api.js";
```
`import api from "../utils/Api.js";`

Please, delete it, and pass it into the constructor of `Card` from `index.js` if you want to use it inside the class


From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,67 +1,128 @@
+
+    if (updatedCard.isLiked !== undefined) {
+      card._isLiked = updatedCard.isLiked;
+      if (card._isLiked) {
+        card._likeButton.classList.add("card__like-button_active");
+      } else {
+        card._likeButton.classList.remove("card__like-button_active");
```
Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,67 +1,128 @@
+      updatedCard = await api.likeCard(card._id);
+    }
+
+    if (updatedCard.isLiked !== undefined) {
```
**Could be improved** 

To have access to any method of the class in `index.js` you can pass `this` to the call:

```js

        this._likeButton.addEventListener('click', () => {
            this._handleLike(this);
        });


```

And now in `index.js` you can call the methods like this:


```js


function addLike(cardInstance) {
        api.likeCard(cardInstance.getId())
            .then((res) => {
               cardInstance.addLike(res)
            })
            .catch((err) => {
                console.error(err);
            });
}


```

The same thing is with deleting cards

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,67 +1,128 @@
+
+    const cardData = await api.addCard(inputValues);
+    const cardElement = createCard(cardData, currentUserId);
+    cardSection.addItem(cardElement);
```
* New cards should be added only with `prepend` to the start of the section, otherwise the user doesn’t see the new card (if there are **6 or more** cards on the page) and may think that it hasn’t been created


Please, pass `true`

`cardSection.addItem(cardElement, true);`

gennady-bars commented on src/pages/index.js:
```diff
@@ -75,51 +136,95 @@ const profileAddPopup = new PopupWithForm(
     renderer: (item) => {
-      const cardElement = createCard(item);
+      const cardElement = createCard(item, currentUserId);
       cardSection.addItem(cardElement, true);
```
` true` is not needed here because the initial cards should be added with `append`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,67 +1,128 @@
+    console.log("Adding card with values:", inputValues);
+
+    const cardData = await api.addCard(inputValues);
+    const cardElement = createCard(cardData, currentUserId);
```
* You should **clear the inputs** after a successful adding of a new card to let the user add the 2nd card again without having to remove the old data manually. 


Please, reset the form here

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,67 +1,128 @@
+  async function makeRequest() {
+    console.log("Updating avatar with values:", inputValues);
+    const userData = await api.updateAvatar(inputValues.link);
+    userInfo.setUserInfo(userData);
```
* You should **clear the inputs** after a successful adding of a new card to let the user add the 2nd card again without having to remove the old data manually. 


Please, reset the form here

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -8,26 +8,44 @@ export default class PopupWithForm extends Popup {
+    this._popupForm.addEventListener("submit", (evt) => {
       evt.preventDefault();
+      const result = this._handleFormSubmit(this._getInputValues());
+      if (result && typeof result.then === "function") {
+        result.then(() => this.close()).catch(console.error);
+      } else {
+        console.error("handleFormSubmit did not return a promise");
```
You don't need it because you have everything in `index.js`

You'll have only this

```js


  setEventListeners() {
    super.setEventListeners();
    this._popupForm.addEventListener("submit", (evt) => {
      evt.preventDefault();
      this._handleFormSubmit(this._getInputValues());
    });
  }

```

gennady-bars reviewed (commented):
Everything is almost fine, but:

* Please, always send this link for your pull request https://github.com/jackmalzone/se_project_aroundtheus/pull/4 
* From the 8th sprint it's not allowed to import anything to the classes. Only class `Popup` can be imported for extending. **`Loose coupling`** should be used: it means that you need to pass callbacks to the constructor rather than using outside elements or functions.
* New cards should be added only with `prepend` to the start of the section, otherwise the user doesn’t see the new card (if there are **6 or more** cards on the page) and may think that it hasn’t been created
* You should **clear the inputs** after a successful adding of a new card to let the user add the 2nd card again without having to remove the old data manually. 
* You need to **reset** the form only  in block `then` in order not to do that immediately after `submit` because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.
* Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

jackmalzone committed:
Refactor and debug application, enhance styling

Refactored Card and UserInfo classes to follow loose coupling and SRP
Encapsulated like/unlike logic within Card class
Fixed add card functionality and form reset after submission
Ensured prepend for new cards to improve user visibility
Added loading screen with rotating planet animation and dynamic travel messages
Enhanced styling for loading screen with circular gradient shadow around planet
Improved overall error handling and logging

jackmalzone commented on src/blocks/page.css:
```diff
@@ -1,9 +1,62 @@
+/* @keyframes gradientShift {
```
got rid of it, not enough time for it to matter

jackmalzone committed:
update readme.md
